File: tests/test_stats.cpp
//
// Created by uki on 11/24/23.
//
// Some tests make use of this dataset from Kaggle:
// "https://www.kaggle.com/datasets/thedevastator/higher-education-predictors-of-student-retention"
//
#include <stats/stats.hpp>
#include <iostream>
#include <gtest/gtest.h>


typedef std::pair<oop::stats::DataFrame, std::optional<std::vector<std::string>>> dataframe_col;
class DataFrameTest : public testing::Test {
protected:
    // Set up all static members when the first test is run.
    static void SetUpTestSuite() {
        if (d2_ == nullptr and d3_ == nullptr) {
            d2_ = new dataframe_col;
            d2_->second = oop::stats::load(filepath, d2_->first, true);
            d3_ = new dataframe_col;
            d3_->second = oop::stats::load(filepath, d2_->first, true);
        }
    }
    // Remember to delete all the pointers when we're done.
    static void TearDownTestSuite() {
        delete d2_;
        d2_ = nullptr;
        delete d3_;
        d3_ = nullptr;
    }
    oop::stats::DataFrame d0_;
    oop::stats::DataFrame d1_{{{0, 1.5, 4}, {3, 14.99, 50}, {10, 8.12, 33}},
                              {{"tc_1", "tc_2", "tc_3"}}};
    static const inline std::string filepath = "./tests/dataset.csv";
    static dataframe_col* d2_; // DataFrame with CSV data
    static dataframe_col* d3_; // DataFrame with partial CSV data
};
dataframe_col* DataFrameTest::d2_ = nullptr;
dataframe_col* DataFrameTest::d3_ = nullptr;

TEST_F(DataFrameTest, LoadCSV) {
    // Load the dataset into our dataframe
    std::vector<std::string> col_types;
    oop::stats::load(filepath, d0_, true);
}

TEST_F(DataFrameTest, LoadCSVPartially) {
    std::vector<int> cols;
    cols.push_back(2);
    oop::stats::load(filepath, d3_->first, true, cols);
}

TEST_F(DataFrameTest, LoadCSVHeader) {
    std::vector<std::string> correct_header{
        "\xEF\xBB\xBFMarital status","Application mode","Application order","Course",
        "Daytime/evening attendance","Previous qualification","Nationality",
        "Mother's qualification","Father's qualification","Mother's occupation",
        "Father's occupation","Displaced","Educational special needs","Debtor",
        "Tuition fees up to date","Gender","Scholarship holder","Age at enrollment",
        "International","Curricular units 1st sem (credited)",
        "Curricular units 1st sem (enrolled)","Curricular units 1st sem (evaluations)",
        "Curricular units 1st sem (approved)","Curricular units 1st sem (grade)",
        "Curricular units 1st sem (without evaluations)",
        "Curricular units 2nd sem (credited)","Curricular units 2nd sem (enrolled)",
        "Curricular units 2nd sem (evaluations)","Curricular units 2nd sem (approved)",
        "Curricular units 2nd sem (grade)","Curricular units 2nd sem (without evaluations)",
        "Unemployment rate","Inflation rate","GDP", "Target"
    };
    ASSERT_TRUE(d2_->first.column_labels()) << "Header not found in CSV file.";
    EXPECT_EQ(*(d2_->first.column_labels()), correct_header) << "Loaded header is incorrect";
}

TEST_F(DataFrameTest, LoadCSVTypesAutodetect) {
    std::vector<std::string> correct_types{
        "long long", "long long", "long long", "long long", "long long",
        "long long", "long long", "long long", "long long", "long long",
        "long long", "long long", "long long", "long long", "long long",
        "long long", "long long", "long long", "long long", "long long",
        "long long", "long long", "long long", "float", "long long", "long long",
        "long long", "long long", "long long", "float", "long long", "float",
        "float", "float", "Catagorical (arma::uword)"};
    EXPECT_EQ(*d2_->second, correct_types);
}

TEST_F(DataFrameTest, LoadCSVCatMap) {
    std::unordered_map<arma::uword, std::string> correct_catmap{
            {0, "Dropout"}, {1, "Graduate"}, {2, "Enrolled"}
    };
    std::unordered_map<arma::uword, std::string> map = d2_->first.get_map(34);
    EXPECT_EQ(map.size(), 3);
    for (std::size_t i = 0; i < 3; ++i) {
        EXPECT_TRUE(map[i]=="Dropout" or map[i]=="Graduate" or map[i]=="Enrolled");
    }
}

TEST_F(DataFrameTest, LoadCSVCorrectShape) {
    EXPECT_EQ(d2_->first.shape().first, 998) << "Wrong number of rows returned by .shape()";
    EXPECT_EQ(d2_->first.shape().second, 35) << "Wrong number of columns returned by .shape()";
}

TEST_F(DataFrameTest, Summarize) {
    std::string correct_output{"tc_1, tc_2, tc_3, \n0 1.5 4 \n3 14.99 50 \n10 8.12 33 \nDataset shape: (3, 3)\n"};
    testing::internal::CaptureStdout();
    d1_.summarize();
    std::string output = testing::internal::GetCapturedStdout();
    EXPECT_EQ(output, correct_output);
}

TEST_F(DataFrameTest, Indexing) {
    auto col1 = std::get<arma::icolvec>(d1_[0]);
    auto col2 = std::get<arma::colvec>(d1_[1]);
    auto col3 = std::get<arma::icolvec>(d1_[2]);

    arma::Col<long long> correct_col1{0, 3, 10};
    arma::Col<double> correct_col2{1.5, 14.99, 8.12};
    arma::Col<long long> correct_col3{4, 50, 33};
    std::string message{"Indexed column has the wrong values."};
    EXPECT_TRUE(arma::approx_equal(col1, correct_col1, "absdiff",
                                   0)) << message;
    EXPECT_TRUE(arma::approx_equal(col2, correct_col2, "absdiff",
                                   0.001)) << message;
    EXPECT_TRUE(arma::approx_equal(col3, correct_col3, "absdiff",
                                   0)) << message;
}

TEST_F(DataFrameTest, AppendColumn) {
    oop::stats::sup_col_types new_column{arma::vec{5.2, 5.5, 6.4}};
    d1_.append_column(new_column, "New Column");
    ASSERT_TRUE(d1_.shape().second == 4);
    ASSERT_TRUE(d1_.shape().first == 3);
    ASSERT_EQ((*d1_.column_labels())[(*d1_.column_labels()).size()-1],
              "New Column");
}

TEST_F(DataFrameTest, AppendInvalidColumn) {
    oop::stats::sup_col_types invalid_column{arma::vec{3.3}};
    ASSERT_ANY_THROW(d1_.append_column(invalid_column));
}

TEST_F(DataFrameTest, InsertColumn) {
    const std::string new_column_label{"New Column"};
    const arma::vec new_column{5.2, 5.5, 6.4};
    oop::stats::sup_col_types new_column_var{new_column};
    d1_.insert_column(new_column_var, 0, new_column_label);
    EXPECT_TRUE(d1_.shape().second == 4);
    EXPECT_TRUE(d1_.shape().first == 3);
    EXPECT_EQ((*d1_.column_labels())[0], new_column_label);
    auto col = std::get<arma::vec>(d1_[0]);
    EXPECT_TRUE(arma::approx_equal(col, new_column, "absdiff", 0));
    EXPECT_NO_THROW(d1_.insert_column(new_column_var, 0));
}

TEST_F(DataFrameTest, InsertInvalidColumn) {
    oop::stats::sup_col_types invalid_column{arma::vec{3.3}};
    ASSERT_ANY_THROW(d1_.insert_column(invalid_column, 0));
}

TEST_F(DataFrameTest, AppendRow) {
    std::vector<oop::stats::sup_single_types> new_row{5, 5.5, 6};

    ASSERT_NO_THROW(d1_.append_row(new_row));
    EXPECT_TRUE(d1_.shape().first == 4);
    EXPECT_TRUE(d1_.shape().second == 3);
}

TEST_F(DataFrameTest, AppendInvalidRowType) {
    std::vector<oop::stats::sup_single_types> invalid_row{5.9, 2.2, 66.4};
    EXPECT_ANY_THROW(d1_.append_row(invalid_row));
}

TEST_F(DataFrameTest, AppendInvalidRowLen) {
    std::vector<oop::stats::sup_single_types> invalid_row{9, 99.9, 5, 44};
    EXPECT_ANY_THROW(d1_.append_row(invalid_row));
}

TEST_F(DataFrameTest, SetValidLabel) {
    std::vector<std::string> new_labels{"l1", "l2", "l3"};
    EXPECT_NO_THROW(d1_.set_column_labels(new_labels));
    EXPECT_EQ(d1_.column_labels(), new_labels) << "Column labels were set incorrectly.";
}

TEST_F(DataFrameTest, SetInvalidLabel) {
    std::vector<std::string> new_labels{"l1", "l2", "l3", "l4"};
    ASSERT_ANY_THROW(d1_.set_column_labels(new_labels));
}

TEST_F(DataFrameTest, CorrectShape) {
    EXPECT_EQ(d1_.shape().first, 3) << "Wrong number of rows returned by .shape()";
    EXPECT_EQ(d1_.shape().second, 3) << "Wrong number of columns returned by .shape()";
}

File: tests/test_stats.py
import numoop
from numoop.stats import DataFrame
import numpy as np
import numpy.typing as npt
import pytest
import pickle
import matplotlib.testing
import matplotlib.testing.compare
import tempfile


class TestDataFrame:
    PLOT_DIR = "./tests/correct_plots"

    def test_init(self) -> None:
        # Init without anything
        df = DataFrame()
        assert df.shape == (0, 0)
        # Init with arguments
        df = DataFrame([[0, 2.0, 100], [10, 11.11, 12]])
        assert df.shape == (2, 3)
        assert np.all((df[0] == np.array([[0], [10]])))

    def test_column_labels(self) -> None:
        labels = ["col1", "col2", "col3"]
        # Init with labels should work as expected.
        df = DataFrame([[0, 2.0, 100],
                               [10, 11.11, 12]], labels)
        assert df.column_labels == labels
        assert np.all(df("col1") == np.array([[0], [10]]))
        other_labels = ["c1", "c2", "c3"]
        # Test label setter
        df.column_labels = other_labels
        assert np.all(df("c3") == np.array([[100], [12]]))
        # Try testing incorrect number of labels
        with pytest.raises(ValueError):
            df.column_labels = ["c1", "c2", "c3", "c4"]
        with pytest.raises(ValueError):
            df.column_labels = ["c1", "c2"]

    def test_append_row(self) -> None:
        df = DataFrame()
        # Appending to an empty DataFrme
        df.append_row([5.5, 1, 4, 7.7])
        # Make sure the values were set correctly
        assert np.isclose(df[0][0], 5.5) and np.isclose(df[3][0], 7.7)
        assert df[1][0] == 1 and df[2][0] == 4
        # Now that we have a row, appending incorrect types or row size should
        # give errors.
        with pytest.raises(ValueError):
            df.append_row([3, 3, 4.5, 7.7])
        with pytest.raises(ValueError):
            df.append_row([1., 2, 3, 4., 5])
        # Make sure the shape is still correct
        assert df.shape == (1, 4)
        # Try appending a second row, check values again.
        df.append_row([5.6, 2, 5, 8.8])
        assert df.shape == (2, 4)
        assert np.isclose(df[0], np.array([[5.5], [5.6]])).all()
        df_2 = DataFrame()
        # We should also be able to use a numpy array to add rows.
        np_row: npt.NDArray[np.int_] = np.array([1, 2, 3, 4])
        df_2.append_row(np_row)
        assert (df_2[0][0] == 1 and df_2[1][0] == 2 and df_2[2][0] == 3 and
                df_2[3][0] == 4)

    def test_insert_row(self) -> None:
        df = DataFrame()
        # Insert a row into an empty DataFrame
        df.insert_row([5, 1.3, 4.4, 7], 0)

        # Inserting incorrect types should raise an error.
        with pytest.raises(ValueError):
            df.insert_row([1, 2.3, 3, 4], 1)
        # Inserting an incorrectly sized row should give an error.
        with pytest.raises(ValueError):
            df.insert_row([1, 2.3, 3.4, 4, 5], 1)
        # Inserting a row out of bounds should also give an error.
        with pytest.raises(IndexError):
            df.insert_row([1, 2.3, 3.4, 4], 100)
        # Make sure the shape is correct.
        assert df.shape == (1, 4)
        # Insert a second row and do checks again
        df.insert_row([6, 2.0, 5.3, 8], 0)
        assert df.shape == (2, 4)
        assert np.all(df[0] == np.array([6, 5])[:, None])

    def test_append_column(self) -> None:
        df = DataFrame()
        # Append a column to an empty DataFrame.
        df.append_column(np.array([[5], [1], [4], [7]]))
        # Appending column with an incorrect number of rows.
        with pytest.raises(ValueError):
            df.append_column(np.array([[1], [2], [3], [4], [5]]))
        # Make sure shape is correct.
        assert df.shape == (4, 1)
        # Append second column and check values and shape.
        df.append_column(np.array([[5.6], [2.0], [5.3], [8.8]]))
        assert df.shape == (4, 2)
        assert np.all(df[0] == np.array([[5], [1], [4], [7]]))

    def test_insert_column(self) -> None:
        df = DataFrame()
        # Insert column into an empty DataFrame.
        df.insert_column(np.array([[5], [1], [4], [7]]), 0)
        # Inserting an incorrect number of rows gives an error.
        with pytest.raises(ValueError):
            df.insert_column(np.array([[1], [2], [3], [4], [5]]), 1)
        # Inserting out of bounds raises an error.
        with pytest.raises(IndexError):
            df.insert_column(np.array([[1], [2], [3], [4]]), 100)
        # Check shape of the DataFrame.
        assert df.shape == (4, 1)
        # Insert second column and check values/shape.
        df.insert_column(np.array([[5.6], [2.0], [5.3], [8.8]]), 0)
        assert df.shape == (4, 2)
        assert np.all(df[1] == np.array([[5], [1], [4], [7]]))

    def test_indexing(self) -> None:
        df = DataFrame([[5, 12.6, 4, 65.2],
                        [1, 55.0, 7, 99.9]])
        # Make sure indexing gives us what we expect.
        assert np.all(df[0] == np.array([[5], [1]]))
        # Take a column out of the DataFrame, change a value, and then reinsert
        # it.
        col = df[0]
        col[1, 0] = 10
        df[0] = col
        assert df[0][1, 0] == 10
        # Trying to set a column with an incorrect number of rows gives an
        # error.
        arr = np.zeros((100, 1))
        with pytest.raises(ValueError):
            df[0] = arr

    def test_dtypes(self) -> None:
        df = DataFrame()

        # Now lets make sure that no hidden type conversions are happening.
        int_ = np.array([[21], [100]], dtype=np.intc)
        long_int = np.array([[21], [100]], dtype=np.int_)
        long_long_int = np.array([[20001], [10]], dtype=np.longlong)
        single = np.array([[22.2], [55.5]], dtype=np.single)
        double = np.array([[23.002], [55.504]], dtype=np.double)
        cx_single = np.array([[33.3 + 10j], [55.5 + 200j]], dtype=np.csingle)
        cx_double = np.array([[44.3 + 100j], [604.2 + 67j]], dtype=np.cdouble)

        cols: list[numoop.types.sup_col_arr_types] = [
            int_, long_int, long_long_int, single, double, cx_single, cx_double
        ]

        list(map(df.append_column, cols))
        assert all([df[i].dtype == cols[i].dtype for i in range(len(cols))])

        # Inserting an invalid type should raise an error.
        bool_ = np.array([[True], [False]], dtype=np.bool_)
        with pytest.raises(TypeError):
            df.append_column(bool_)  # type: ignore

    def test_view(self) -> None:
        df = DataFrame([[5, 12.6, 4, 65.2],
                        [1, 55.0, 7, 99.9]],
                        labels=["col1", "col2", "col3", "col4"])
        # Make sure the view has the correct values.
        assert np.all(df.view(0) == np.array([[5], [1]]))
        # We should also be able to access a column with it's label:
        assert np.all(df.view("col3") == np.array([[4], [7]]))

        # Editing a view is not allowed.
        with pytest.raises(ValueError):
            df.view(0)[1] = 10

    def test_pickle(self) -> None:
        labs = ["col1", "col2", "col3", "col4"]
        df = DataFrame([[0, 12.6, 4, 65.2],
                        [1, 55.0, 7, 99.9]],
                              labs)
        cat_map = {0: "cat1", 1: "cat2"}
        df.set_map(0, cat_map)
        # Pickle data
        data = pickle.dumps(df, -1)
        # Unplickle data
        df_unpickled = pickle.loads(data)
        # Make sure all unpickled data is correct.
        assert np.all(df[0] == df_unpickled[0])
        assert np.all(np.isclose(df[1], df_unpickled[1]))
        assert np.all(df[2] == df_unpickled[2])
        assert np.all(np.isclose(df[3], df_unpickled[3]))
        assert df_unpickled.column_labels == labs
        assert df_unpickled.get_map(0) == cat_map

    def test_set_map(self) -> None:
        df = DataFrame([[1, 12.6, 0],
                        [1, 55.0, 2],
                        [0, 33.1, 3]])
        cat_map_one = {0: "cat1", 1: "cat2"}
        cat_map_two = {0: "cat3", 1: "cat4", 2: "cat5", 3: "cat6"}

        df.set_map(0, cat_map_one)
        assert df.get_map(0) == cat_map_one

        with pytest.raises(IndexError):
            df.set_map(20, cat_map_two)

        df.set_map(2, cat_map_two)
        assert df.get_map(2) == cat_map_two
        assert df.get_map(0) == cat_map_one

    def test_plots(self) -> None:
        df = DataFrame([[1, 12.6, 0],
                        [1, 55.0, 2],
                        [0, 33.1, 3],
                        [0, 50.5, 3],
                        [0, 22.2, 0]])
        df.set_map(0, {0: "cat1", 1: "cat2"})
        df.column_labels = ["First Column", "Second Column", "Third Column"]

        plot_args = [(1, "hist"), (0, "bar"), (2, "bar"), (1, "line"),
                     (0, "pie")]

        # Improve reproducibility
        matplotlib.testing.setup()
        with tempfile.TemporaryDirectory() as f:
            for file_no, (idx, plot_type) in enumerate(plot_args):
                fig, ax = df.make_plot(idx, plot_type)
                file_name = f"fig{str(file_no)}.png"
                temp_fig_path = f"{f}/{file_name}"
                true_fig_path = f"{self.PLOT_DIR}/{file_name}"
                fig.savefig(temp_fig_path)
                assert matplotlib.testing.compare.compare_images(
                    true_fig_path, temp_fig_path, 0.001
                ) is None

        invalid_calls = [(1, "pie"), (1, "bar")]
        for idx, plot_type in invalid_calls:
            with pytest.raises(TypeError):
                df.make_plot(idx, plot_type)

        with pytest.raises(AttributeError):
            df.make_plot(0, "invalid_plot_type")


class TestLoadCSV:
    CSV_PATH = "./tests/dataset.csv"
    CSV_CORRECT_LABELS = [
        '\ufeffMarital status', 'Application mode', 'Application order',
        'Course', 'Daytime/evening attendance', 'Previous qualification',
        'Nationality', "Mother's qualification", "Father's qualification",
        "Mother's occupation", "Father's occupation", 'Displaced',
        'Educational special needs', 'Debtor', 'Tuition fees up to date',
        'Gender', 'Scholarship holder', 'Age at enrollment', 'International',
        'Curricular units 1st sem (credited)',
        'Curricular units 1st sem (enrolled)',
        'Curricular units 1st sem (evaluations)',
        'Curricular units 1st sem (approved)',
        'Curricular units 1st sem (grade)',
        'Curricular units 1st sem (without evaluations)',
        'Curricular units 2nd sem (credited)',
        'Curricular units 2nd sem (enrolled)',
        'Curricular units 2nd sem (evaluations)',
        'Curricular units 2nd sem (approved)',
        'Curricular units 2nd sem (grade)',
        'Curricular units 2nd sem (without evaluations)',
        'Unemployment rate', 'Inflation rate', 'GDP', 'Target']

    def test_full_load(self) -> None:
        df = numoop.stats.load(self.CSV_PATH, header=True)
        assert df.column_labels == self.CSV_CORRECT_LABELS
        first_col = df[0]
        assert df.shape == (499, 35)
        assert first_col.shape[0] == 499
        assert first_col.dtype.name == 'int64'
        assert df[df.shape[1]-2].dtype.name == 'float32'

        catmap = df.get_map(df.shape[1]-1)
        assert isinstance(catmap, dict)
        assert catmap[0] == "Dropout"
        assert catmap[1] == "Graduate"
        assert catmap[2] == "Enrolled"

        # We should also be able to load into a DataFrame
        df = DataFrame()
        numoop.stats.load(self.CSV_PATH, df, True)

    def test_partial_load(self) -> None:
        df = numoop.stats.load(self.CSV_PATH, header=True, columns=[1, 6])
        assert df.column_labels == [self.CSV_CORRECT_LABELS[1],
                                    self.CSV_CORRECT_LABELS[6]]
        first_col = df[0]
        assert df.shape == (499, 2)
        assert first_col.shape[0] == 499
        assert first_col.dtype.name == 'int64'
        assert df[df.shape[1]-2].dtype.name == 'int64'
